source/state_representation/src/trajectory/CartesianTrajectory.cpp

```diff
@@ -0,0 +1,210 @@
+  this->add_point(point, duration);
+}
+
+CartesianTrajectory::CartesianTrajectory(
+    const std::string& name, const std::vector<CartesianState>& points,
+    const std::vector<std::chrono::nanoseconds>& durations
+)
+    : TrajectoryBase<CartesianTrajectoryPoint>(name) {
+  this->set_type(StateType::CARTESIAN_TRAJECTORY);
+  if (points.empty()) {
+    throw exceptions::IncompatibleSizeException("No points provided");
+  } else if (std::ranges::any_of(points, [&](const auto& p) { return p.is_empty(); })) {
+    throw exceptions::EmptyStateException("Vector contains at least one point that is empty");
+  } else if (!std::ranges::all_of(points, [&](const auto& p) {
+               return p.get_reference_frame() == points.front().get_reference_frame();
+             })) {
+    throw exceptions::IncompatibleReferenceFramesException("Incompatible reference frames within the points vector");
+  }
+  this->reference_frame_ = points[0].get_reference_frame();
+  try {
+    this->add_points(points, durations);
+  } catch (...) {
+    throw;
+  }
+}
+
+const std::string& CartesianTrajectory::get_reference_frame() const {
+  return this->reference_frame_;
+}
+
+void CartesianTrajectory::set_reference_frame(const CartesianPose& pose) {
+  if (pose.is_empty()) {
+    throw exceptions::EmptyStateException("Pose is empty");
+  } else if (this->get_size() == 0) {
+    throw exceptions::EmptyStateException("Trajectory is empty");
+  }
+  this->reference_frame_ = pose.get_reference_frame();
+  auto points = this->get_points();
+  for (auto& point : points) {
+    point *= pose;
+  }
+  try {
+    this->set_points(points, this->get_durations());
+  } catch (...) {
+    throw;
+  }
+}
+
+void CartesianTrajectory::add_point(const CartesianState& new_point, const std::chrono::nanoseconds& duration) {
+  if (new_point.is_empty()) {
+    throw exceptions::EmptyStateException("The Cartesian state provided is empty");
+  } else if (new_point.get_reference_frame() != this->reference_frame_) {
+    throw exceptions::IncompatibleReferenceFramesException(
+        "Incompatible reference frames: " + new_point.get_reference_frame() + " and " + this->reference_frame_
+    );
+  }
+  CartesianTrajectoryPoint trajectory_point;
+  trajectory_point.data = new_point.data();
+  trajectory_point.duration = duration;
+  trajectory_point.name = new_point.get_name();
+  this->TrajectoryBase<CartesianTrajectoryPoint>::add_point(trajectory_point);
+}
+
+void CartesianTrajectory::add_points(
+    const std::vector<CartesianState>& new_points, const std::vector<std::chrono::nanoseconds>& durations
+) {
+  if (new_points.empty()) {
+    throw exceptions::IncompatibleSizeException("No points provided");
+  } else if (new_points.size() != durations.size()) {
+    throw exceptions::IncompatibleSizeException("The size of the points and durations vectors are not equal");
+  } else if (std::ranges::any_of(new_points, [&](const auto& p) { return p.is_empty(); })) {
```

Comment:
While the check is correct, I wonder if it's necessary. `add_point` will do the same checks again, and in `add_point` we even provide the names of the incompatible reference frames in the exception message which I find almost more valuable. I'd prefer to get the message from line 79 that tells me where to look instead of the generic message from line 99 that just says that there is at least one problem. In that regard, I would skip the checks of line 96 and 98 but that's just my opinion